extension/wrapper.cpp

```diff
@@ -144,66 +183,130 @@ Satrec__sgp4(PyObject *self, PyObject *args)
-    {"ecco", T_DOUBLE, O(ecco), READONLY,
+    {"ecco", T_DOUBLE, O(ecco), 0,
      PyDoc_STR("Eccentricity.")},
-    {"argpo", T_DOUBLE, O(argpo), READONLY,
+    {"argpo", T_DOUBLE, O(argpo), 0,
      PyDoc_STR("Argument of perigee in radians.")},
-    {"mo", T_DOUBLE, O(mo), READONLY,
+    {"mo", T_DOUBLE, O(mo), 0,
      PyDoc_STR("Mean anomaly in radians.")},
-    {"no_kozai", T_DOUBLE, O(no_kozai), READONLY,
+    {"no_kozai", T_DOUBLE, O(no_kozai), 0,
      PyDoc_STR("Mean motion in radians per minute.")},
-    {"revnum", T_LONG, O(revnum), READONLY,
+    {"revnum", T_LONG, O(revnum), 0,
      PyDoc_STR("Integer revolution number at the epoch.")},
 
     /* For compatibility with the old struct members, also accept the
        plain name "no". */
 
-    {"no", T_DOUBLE, O(no_kozai), READONLY,
+    {"no", T_DOUBLE, O(no_kozai), 0,
      PyDoc_STR("Alias for the more carefully named ``no_kozai``.")},
 
     /* Derived values that do not appear explicitly in the TLE. */
 
     {"method", T_CHAR, O(method), READONLY,
      PyDoc_STR("Method, either 'n' near earth or 'd' deep space.")},
+    {"error", T_INT, O(method), 0,
+     PyDoc_STR("Error code (1-6) documented in sgp4()")},
+    {"a", T_DOUBLE, O(a), 0,
+     PyDoc_STR("semi-major axis")},
+    {"altp", T_DOUBLE, O(altp), 0,
+     PyDoc_STR("altitude of perigee")},
+    {"alta", T_DOUBLE, O(alta), 0,
+     PyDoc_STR("altitude of perigee")},
+
+    /* Single averaged mean elements */
+
+    {"am", T_DOUBLE, O(am), 0,
+     PyDoc_STR("am: Average semi-major axis")},
+    {"em", T_DOUBLE, O(em), 0,
+     PyDoc_STR("em: Average eccentricity")},
+    {"im", T_DOUBLE, O(im), 0,
+     PyDoc_STR("im: Average inclination")},
+    {"Om", T_DOUBLE, O(Om), 0,
+     PyDoc_STR("Om: Average right ascension of ascending node")},
+    {"om", T_DOUBLE, O(om), 0,
+     PyDoc_STR("om: Average argument of perigee")},
+    {"mm", T_DOUBLE, O(mm), 0,
+     PyDoc_STR("mm: Average mean anomaly")},
+    {"nm", T_DOUBLE, O(nm), 0,
+     PyDoc_STR("nm: Average mean motion")},
+
+    /* Gravity-constant dependent values (initialized by sgp4init() */
+
+    {"tumin", T_DOUBLE, O(tumin), 0,
+     PyDoc_STR("minutes in one time unit")},
+    {"mu", T_DOUBLE, O(mu), 0,
+     PyDoc_STR("Earth gravitational parameter")},
+    {"radiusearthkm", T_DOUBLE, O(radiusearthkm), 0,
+     PyDoc_STR("radius of the earth in km")},
+    {"xke", T_DOUBLE, O(xke), 0,
+     PyDoc_STR("reciprocal of tumin")},
+    {"j2", T_DOUBLE, O(j2), 0,
+     PyDoc_STR("un-normalized zonal harmonic j2 value")},
+    {"j3", T_DOUBLE, O(j3), 0,
+     PyDoc_STR("un-normalized zonal harmonic j3 value")},
+    {"j4", T_DOUBLE, O(j4), 0,
+     PyDoc_STR("un-normalized zonal harmonic j4 value")},
+    {"j3oj2", T_DOUBLE, O(j3oj2), 0,
+     PyDoc_STR("j3 divided by j2")},
+
+    /* Other convenience variables (some required by propagation.py) */
+    {"t", T_DOUBLE, O(t), 0,
+     PyDoc_STR("Last tsince input to sgp4()")},
```

Comment:
There is an unspoken design principle I had in mind when writing this C interface; let's add an Issue to have me add that to some developer docs. The principle was that, to the extent possible, we should take a C++ library that follows the "call-method-and-mutate" approach that's similar to Python's own `list.sort()` and to how PyEphem works, and that instead presents a modern functional interface where object attributes are generally immutable and where methods return information rather than leaving the object with visibly mutated state, more like the modern `sorted()` routine that has replaced `list.sort()` in modern Python code.

Is there a use case for exposing `t`? For each value below, I'm interested in which ones stay static for the lifetime of the object, and which ones change as the satellite is given different `tsince` values.

Hmm.

All of which raises a question I hadn't even thought of until this point: that having `sgp4init()` as a method — like I asked when you'd so nicely wrapped it up as a constructor originally — that mutates the satellite object over its lifetime into a series of different satellite orbits, a new one each time it's called, kind of contradicts how I'm asking for attributes to be handled where they'd ideally be unchanging and read-only.

I think the big question that will resolve my inconsistency here is: what are all the use cases you are working to cover with the addition of sgp4init() and the exposure of additional attributes? Happily, our need to add those use cases to the documentation means that you won't be wasting effort to write up the use cases; that work will feed directly into the big docstring that gets sent up to PyPI.

Feel free to not write the documentation text yourself, as I'm happy to take on that burden. If you'd like, just add a little `newdocs.md` file at the root of this PR that says things like `For reason <X> folks in situation <Y> need the grav parameters to do <Z>, so:` and then shows a small complete script that demonstrates to users how to accomplish it. I can then go through them, transfer them to the main docstring, and write bigger explanations once this PR is merged.

And, as I read through them, I'll understand the use cases covered by these new features, and will be in a better position to decide on whether I made a mistake by suggesting that `sgp4init()` should be a plain method.

I wonder if my thought was "only advanced users would ever call `sgp4init()` so why not have it update the state directly?"

Reply:
Documentation is nice, more especially for advanced use cases.

In essence, my application is repeatedly reinitializing a satrec instance with tweaked elements to evaluate its affect on RMS errors when compared with observations.  Being able to work with the raw orbital elements, rather than having to re-construct a TLE to then re-import has significant ease-of-use as well as performance considerations.

Reply:
Thanks for the details! How often are you re-submitting the whole slate of elements, versus tweaking just one element and leaving the others static (since the latter case would be a very strong case for a writable element, since calling `sgp4init()` all over again would uselessly set all the elements over again to the values they already have)?